### ISSUE TITLE
"Dataset does not exist" in incremental receives in current master

### ISSUE THREAD

**Issue by zrav** [zfsonlinux/zfs#8067]
### System information

Type | Version/Name
 --- | --- 
Distribution Name	| Ubuntu
Distribution Version	| 18.10
Linux Kernel	| 4.18.0
Architecture	| AMD64
ZFS Version	| 0.7.9 (source), current master (target)
SPL Version	| 0.7.9 (source), current master (target)

### Describe the problem you're observing
After migrating from 0.7.5 to 0.7.9 and enabling the new pool features (large_dnode and userobj_accounting active), we ran into problems, one being receive performance issues, the other that lots of sends were failing with "cannot receive incremental stream: kernel modules must be upgraded to receive this stream.", "cannot receive incremental stream: invalid backup stream" and "cannot receive incremental stream: dataset does not exist". Since 0.7.10 fixed multiple receive issues, I tried upgrading the receiver to 0.7.11. It did seem to fix most failures but I'm unable to confirm, as 0.7.11 exhibits the same performance issues as 0.7.9. So I tried compiling current master (b74f48f). Most of the occurrences of failed receives went away, but some instances of "dataset does not exist" persist. The receive performance issues also didn't go away completely, but I'm opening a separate issue for that.

The source pool holds KVM images and LXD containers, but so far only LXD container datasets are affected. All snapshots from before the new pool features were enabled can be received. On the affected datasets some later snapshots reproducibly trigger the error. The failure doesn't necessarily occur right at the start of the stream, but can be dozens of MB in.
### Describe how to reproduce the problem
The source is a production server running Ubuntu 18.10 with 0.7.9 on a mirrored NVMe pool.

```
root@source:~# modinfo zfs
filename:       /lib/modules/4.18.0-10-generic/kernel/zfs/zfs/zfs.ko
version:        0.7.9-3ubuntu6
license:        CDDL
author:         OpenZFS on Linux
description:    ZFS
srcversion:     F82D0AFD503F68490D1E9A4
depends:        spl,znvpair,zcommon,zunicode,zavl,icp
retpoline:      Y
name:           zfs
vermagic:       4.18.0-10-generic SMP mod_unload
...

root@source:~# zpool status
  pool: tank
 state: ONLINE
  scan: scrub repaired 0B in 0h20m with 0 errors on Sun Oct 14 00:44:28 2018
config:

        NAME                                             STATE     READ WRITE CKSUM
        tank                                             ONLINE       0     0     0
          mirror-0                                       ONLINE       0     0     0
            nvme-Micron_9200_MTFDHAL3T8TCT_XXXXXXXXXXXX  ONLINE       0     0     0
            nvme-Micron_9200_MTFDHAL3T8TCT_XXXXXXXXXXXX  ONLINE       0     0     0

errors: No known data errors

root@source:~# zpool get all tank
NAME  PROPERTY                       VALUE                          SOURCE
tank  size                           3.47T                          -
tank  capacity                       45%                            -
tank  altroot                        -                              default
tank  health                         ONLINE                         -
tank  guid                           17599409880014544006           -
tank  version                        -                              default
tank  bootfs                         -                              default
tank  delegation                     on                             default
tank  autoreplace                    off                            default
tank  cachefile                      -                              default
tank  failmode                       wait                           default
tank  listsnapshots                  off                            default
tank  autoexpand                     off                            default
tank  dedupditto                     0                              default
tank  dedupratio                     1.00x                          -
tank  free                           1.89T                          -
tank  allocated                      1.57T                          -
tank  readonly                       off                            -
tank  ashift                         13                             local
tank  comment                        -                              default
tank  expandsize                     -                              -
tank  freeing                        0                              -
tank  fragmentation                  23%                            -
tank  leaked                         0                              -
tank  multihost                      off                            default
tank  feature@async_destroy          enabled                        local
tank  feature@empty_bpobj            active                         local
tank  feature@lz4_compress           active                         local
tank  feature@multi_vdev_crash_dump  enabled                        local
tank  feature@spacemap_histogram     active                         local
tank  feature@enabled_txg            active                         local
tank  feature@hole_birth             active                         local
tank  feature@extensible_dataset     active                         local
tank  feature@embedded_data          active                         local
tank  feature@bookmarks              enabled                        local
tank  feature@filesystem_limits      enabled                        local
tank  feature@large_blocks           enabled                        local
tank  feature@large_dnode            active                         local
tank  feature@sha512                 enabled                        local
tank  feature@skein                  enabled                        local
tank  feature@edonr                  enabled                        local
tank  feature@userobj_accounting     active                         local

root@source:~# zfs get all tank/lxd/containers/mycontainer
NAME                             PROPERTY              VALUE                                                                   SOURCE
tank/lxd/containers/mycontainer  type                  filesystem                                                              -
tank/lxd/containers/mycontainer  creation              Thu Aug 16 11:48 2018                                                   -
tank/lxd/containers/mycontainer  used                  112G                                                                    -
tank/lxd/containers/mycontainer  available             11.2G                                                                   -
tank/lxd/containers/mycontainer  referenced            18.6G                                                                   -
tank/lxd/containers/mycontainer  compressratio         1.16x                                                                   -
tank/lxd/containers/mycontainer  mounted               no                                                                      -
tank/lxd/containers/mycontainer  origin                tank/lxd/containers/myorigin@snapshot-snap0                             -
tank/lxd/containers/mycontainer  quota                 none                                                                    default
tank/lxd/containers/mycontainer  reservation           none                                                                    default
tank/lxd/containers/mycontainer  recordsize            128K                                                                    default
tank/lxd/containers/mycontainer  mountpoint            /var/snap/lxd/common/lxd/storage-pools/default/containers/mycontainer   local
tank/lxd/containers/mycontainer  sharenfs              off                                                                     default
tank/lxd/containers/mycontainer  checksum              on                                                                      default
tank/lxd/containers/mycontainer  compression           lz4                                                                     inherited from tank
tank/lxd/containers/mycontainer  atime                 off                                                                     inherited from tank
tank/lxd/containers/mycontainer  devices               on                                                                      inherited from tank/lxd
tank/lxd/containers/mycontainer  exec                  on                                                                      inherited from tank/lxd
tank/lxd/containers/mycontainer  setuid                on                                                                      inherited from tank/lxd
tank/lxd/containers/mycontainer  readonly              off                                                                     default
tank/lxd/containers/mycontainer  zoned                 off                                                                     default
tank/lxd/containers/mycontainer  snapdir               hidden                                                                  default
tank/lxd/containers/mycontainer  aclinherit            passthrough                                                             inherited from tank
tank/lxd/containers/mycontainer  createtxg             291315                                                                  -
tank/lxd/containers/mycontainer  canmount              noauto                                                                  local
tank/lxd/containers/mycontainer  xattr                 sa                                                                      inherited from tank
tank/lxd/containers/mycontainer  copies                1                                                                       default
tank/lxd/containers/mycontainer  version               5                                                                       -
tank/lxd/containers/mycontainer  utf8only              off                                                                     -
tank/lxd/containers/mycontainer  normalization         none                                                                    -
tank/lxd/containers/mycontainer  casesensitivity       sensitive                                                               -
tank/lxd/containers/mycontainer  vscan                 off                                                                     default
tank/lxd/containers/mycontainer  nbmand                off                                                                     default
tank/lxd/containers/mycontainer  sharesmb              off                                                                     default
tank/lxd/containers/mycontainer  refquota              29.8G                                                                   local
tank/lxd/containers/mycontainer  refreservation        none                                                                    default
tank/lxd/containers/mycontainer  guid                  8998195550931737089                                                     -
tank/lxd/containers/mycontainer  primarycache          all                                                                     default
tank/lxd/containers/mycontainer  secondarycache        all                                                                     default
tank/lxd/containers/mycontainer  usedbysnapshots       97.1G                                                                   -
tank/lxd/containers/mycontainer  usedbydataset         15.3G                                                                   -
tank/lxd/containers/mycontainer  usedbychildren        0B                                                                      -
tank/lxd/containers/mycontainer  usedbyrefreservation  0B                                                                      -
tank/lxd/containers/mycontainer  logbias               latency                                                                 default
tank/lxd/containers/mycontainer  dedup                 off                                                                     default
tank/lxd/containers/mycontainer  mlslabel              none                                                                    default
tank/lxd/containers/mycontainer  sync                  disabled                                                                inherited from tank/lxd
tank/lxd/containers/mycontainer  dnodesize             auto                                                                    inherited from tank
tank/lxd/containers/mycontainer  refcompressratio      1.21x                                                                   -
tank/lxd/containers/mycontainer  written               64.5M                                                                   -
tank/lxd/containers/mycontainer  logicalused           110G                                                                    -
tank/lxd/containers/mycontainer  logicalreferenced     18.6G                                                                   -
tank/lxd/containers/mycontainer  volmode               default                                                                 default
tank/lxd/containers/mycontainer  filesystem_limit      none                                                                    default
tank/lxd/containers/mycontainer  snapshot_limit        none                                                                    default
tank/lxd/containers/mycontainer  filesystem_count      none                                                                    default
tank/lxd/containers/mycontainer  snapshot_count        none                                                                    default
tank/lxd/containers/mycontainer  snapdev               hidden                                                                  default
tank/lxd/containers/mycontainer  acltype               posixacl                                                                inherited from tank
tank/lxd/containers/mycontainer  context               none                                                                    default
tank/lxd/containers/mycontainer  fscontext             none                                                                    default
tank/lxd/containers/mycontainer  defcontext            none                                                                    default
tank/lxd/containers/mycontainer  rootcontext           none                                                                    default
tank/lxd/containers/mycontainer  relatime              off                                                                     default
tank/lxd/containers/mycontainer  redundant_metadata    all                                                                     default
tank/lxd/containers/mycontainer  overlay               off                                                                     default
```
The target is an unused server, freshly installed with Ubuntu 18.10 and ZoL compiled from current master (b74f48f).

```
root@target:~# modinfo zfs
filename:       /lib/modules/4.18.0-10-generic/updates/dkms/zfs.ko
version:        0.8.0-rc1_78_gb74f48fe1
license:        CDDL
author:         OpenZFS on Linux
description:    ZFS
alias:          devname:zfs
alias:          char-major-10-249
srcversion:     DC7B3351B6223B82AD2DF43
depends:        zlua,spl,znvpair,zcommon,icp,zunicode,zavl
retpoline:      Y
name:           zfs
vermagic:       4.18.0-10-generic SMP mod_unload

root@target:~# zpool status
  pool: tank
 state: ONLINE
status: Some supported features are not enabled on the pool. The pool can
        still be used, but some features are unavailable.
action: Enable all features using 'zpool upgrade'. Once this is done,
        the pool may no longer be accessible by software that does not support
        the features. See zpool-features(5) for details.
  scan: none requested
config:

        NAME                                           STATE     READ WRITE CKSUM
        tank                                           ONLINE       0     0     0
          mirror-0                                     ONLINE       0     0     0
            ata-WDC_WD1003FBYX-01Y7B0_WD-WCAWXXXXXXXX  ONLINE       0     0     0
            ata-WDC_WD1003FBYX-01Y7B0_WD-WCAWXXXXXXXX  ONLINE       0     0     0

errors: No known data errors

root@target:~# zpool get all tank
NAME  PROPERTY                       VALUE                          SOURCE
tank  size                           928G                           -
tank  capacity                       1%                             -
tank  altroot                        -                              default
tank  health                         ONLINE                         -
tank  guid                           13711680443329247119           -
tank  version                        -                              default
tank  bootfs                         -                              default
tank  delegation                     on                             default
tank  autoreplace                    off                            default
tank  cachefile                      -                              default
tank  failmode                       wait                           default
tank  listsnapshots                  off                            default
tank  autoexpand                     off                            default
tank  dedupditto                     0                              default
tank  dedupratio                     1.00x                          -
tank  free                           917G                           -
tank  allocated                      11.2G                          -
tank  readonly                       off                            -
tank  ashift                         0                              default
tank  comment                        -                              default
tank  expandsize                     16.0E                          -
tank  freeing                        0                              -
tank  fragmentation                  0%                             -
tank  leaked                         0                              -
tank  multihost                      off                            default
tank  checkpoint                     -                              -
tank  load_guid                      9939773913554085201            -
tank  feature@async_destroy          enabled                        local
tank  feature@empty_bpobj            active                         local
tank  feature@lz4_compress           active                         local
tank  feature@multi_vdev_crash_dump  enabled                        local
tank  feature@spacemap_histogram     active                         local
tank  feature@enabled_txg            active                         local
tank  feature@hole_birth             active                         local
tank  feature@extensible_dataset     active                         local
tank  feature@embedded_data          active                         local
tank  feature@bookmarks              enabled                        local
tank  feature@filesystem_limits      enabled                        local
tank  feature@large_blocks           enabled                        local
tank  feature@large_dnode            active                         local
tank  feature@sha512                 enabled                        local
tank  feature@skein                  enabled                        local
tank  feature@edonr                  enabled                        local
tank  feature@userobj_accounting     active                         local
tank  feature@encryption             disabled                       local
tank  feature@project_quota          disabled                       local
tank  feature@device_removal         disabled                       local
tank  feature@obsolete_counts        disabled                       local
tank  feature@zpool_checkpoint       disabled                       local
tank  feature@spacemap_v2            disabled                       local
tank  feature@allocation_classes     disabled                       local
tank  feature@resilver_defer         disabled                       local

root@target:~# zfs get all tank/test
NAME       PROPERTY              VALUE                  SOURCE
tank/test  type                  filesystem             -
tank/test  creation              Sat Oct 27 12:17 2018  -
tank/test  used                  11.2G                  -
tank/test  available             888G                   -
tank/test  referenced            37K                    -
tank/test  compressratio         1.36x                  -
tank/test  mounted               yes                    -
tank/test  quota                 none                   default
tank/test  reservation           none                   default
tank/test  recordsize            128K                   default
tank/test  mountpoint            /tank/test             default
tank/test  sharenfs              off                    default
tank/test  checksum              on                     default
tank/test  compression           lz4                    inherited from tank
tank/test  atime                 off                    inherited from tank
tank/test  devices               on                     default
tank/test  exec                  on                     default
tank/test  setuid                on                     default
tank/test  readonly              off                    default
tank/test  zoned                 off                    default
tank/test  snapdir               hidden                 default
tank/test  aclinherit            passthrough            inherited from tank
tank/test  createtxg             318                    -
tank/test  canmount              on                     default
tank/test  xattr                 sa                     inherited from tank
tank/test  copies                1                      default
tank/test  version               5                      -
tank/test  utf8only              on                     -
tank/test  normalization         formD                  -
tank/test  casesensitivity       mixed                  -
tank/test  vscan                 off                    default
tank/test  nbmand                off                    default
tank/test  sharesmb              off                    default
tank/test  refquota              none                   default
tank/test  refreservation        none                   default
tank/test  guid                  7143646163179245210    -
tank/test  primarycache          all                    default
tank/test  secondarycache        all                    default
tank/test  usedbysnapshots       0B                     -
tank/test  usedbydataset         37K                    -
tank/test  usedbychildren        11.2G                  -
tank/test  usedbyrefreservation  0B                     -
tank/test  logbias               latency                default
tank/test  objsetid              515                    -
tank/test  dedup                 off                    default
tank/test  mlslabel              none                   default
tank/test  sync                  disabled               inherited from tank
tank/test  dnodesize             legacy                 default
tank/test  refcompressratio      1.00x                  -
tank/test  written               37K                    -
tank/test  logicalused           14.9G                  -
tank/test  logicalreferenced     18.5K                  -
tank/test  volmode               default                default
tank/test  filesystem_limit      none                   default
tank/test  snapshot_limit        none                   default
tank/test  filesystem_count      none                   default
tank/test  snapshot_count        none                   default
tank/test  snapdev               hidden                 default
tank/test  acltype               posixacl               inherited from tank
tank/test  context               none                   default
tank/test  fscontext             none                   default
tank/test  defcontext            none                   default
tank/test  rootcontext           none                   default
tank/test  relatime              off                    default
tank/test  redundant_metadata    all                    default
tank/test  overlay               off                    default
tank/test  encryption            off                    default
tank/test  keylocation           none                   default
tank/test  keyformat             none                   default
tank/test  pbkdf2iters           0                      default
tank/test  special_small_blocks  0                      default
```
The source datasets to be transferred:
```
root@source:~# zfs list -t snapshot -r tank/lxd/containers/mycontainer
NAME                                                                     USED  AVAIL  REFER  MOUNTPOINT
tank/lxd/containers/mycontainer@zfs-auto-snap_monthly-2018-09-01-0452  54.0M      -  19.9G  -
...
tank/lxd/containers/mycontainer@zfs-auto-snap_daily-2018-10-23-0425     456M      -  12.6G  -
tank/lxd/containers/mycontainer@zfs-auto-snap_daily-2018-10-24-0425     953M      -  13.7G  -
tank/lxd/containers/mycontainer@zfs-auto-snap_daily-2018-10-25-0425     540M      -  16.8G  -
tank/lxd/containers/mycontainer@zfs-auto-snap_daily-2018-10-26-0425     260M      -  17.8G  -
tank/lxd/containers/mycontainer@zfs-auto-snap_daily-2018-10-27-0425    57.0M      -  18.6G  -
```
Sending initial snapshot and trying to send the remaining incremental snapshots:
```
root@target:~# ssh source "zfs send tank/lxd/containers/mycontainer@zfs-auto-snap_monthly-2018-09-01-0452" | pv | zfs receive tank/test/mycontainer
20.9GiB 0:03:31 [ 101MiB/s]

root@target:~# ssh source "zfs send -I tank/lxd/containers/mycontainer@zfs-auto-snap_monthly-2018-09-01-0452 tank/lxd/containers/mycontainer@zfs-auto-snap_daily-2018-10-27-0425" | pv | zfs receive -F tank/test/mycontainer
cannot receive incremental stream: dataset does not exist
92.4GiB 0:55:03 [28.6MiB/s]

root@target:~# zfs list -t snapshot -r tank/test
NAME                                                           USED  AVAIL  REFER  MOUNTPOINT
tank/test/mycontainer@zfs-auto-snap_monthly-2018-09-01-0452  47.5M      -  15.2G  -
...
tank/test/mycontainer@zfs-auto-snap_daily-2018-10-23-0425     243M      -  9.57G  -
tank/test/mycontainer@zfs-auto-snap_daily-2018-10-24-0425       0B      -  9.95G  -

root@target:~# ssh aource "zfs send -I tank/lxd/containers/mycontainer@zfs-auto-snap_daily-2018-10-24-0425 tank/lxd/containers/mycontainer@zfs-auto-snap_daily-2018-10-27-0425" | pv | zfs receive -F tank/test/mycontainer
cannot receive incremental stream: dataset does not exist
 190MiB 0:00:08 [23.7MiB/s]
```
As I can't hand out these company datasets, I'm currently trying to create a generic container dataset stream that triggers the issue, without luck so far. If someone could point me into the right direction on how to extract useful debug info from the setup, I'd be happy to try.
### Include any warning/errors/backtraces from the system logs
Nothing unusual in dmesg / syslogs.

**Comment by kpande** [comment]
please use [zfs-discuss or IRC](http://list.zfsonlinux.org) for this issue. there are a few debugging options to try before filing an issue, like using zstreamdump to inspect stream contents.

**Comment by zrav** [comment]
Why would you close what is clearly a bug Issue? This reproduces on current master. The only problem I have is that I can't hand out the affected dataset dumps as they contain company data. Once more debugging information is available I will have to post it somewhere, and zfs-discuss or IRC are not the place for that.

**Comment by kpande** [comment]
because there is no valid reproducible case. if one can be provided, reopen it.

**Comment by zrav** [comment]
I have a reproducer that triggers the problem on both 0.7.13 as well as current master (48ed0f9). Can a developer please contact me or let me know how to reach them to arrange transfer of the send streams.
Also please reopen this issue.

**Comment by kpande** [comment]
you can share the reproducer here.

**Comment by zrav** [comment]
No, I can't. The affected dataset contains private data which I can't share with the world.

**Comment by kpande** [comment]
well, that is unfortunate.. I wish you luck!

**Comment by zrav** [comment]
It's not like the situation is new, I've seen private exchanges of reproducers before in this project.
Please reopen the issue, I don't have the rights to do it.

**Comment by behlendorf** [comment]
@zrav sorry to hear you ran in to this, but I should have a fix for this issue.  Since you mentioned you have a reproducer would it be possible for you to verify the proposed fix.  The following branch is based on the latest master and includes the fix.

https://github.com/behlendorf/zfs/tree/zfs-recv-realloc

If you'd rather test using 0.7.13, you would only need to manually apply this hunk from the patch.

```diff
diff --git a/module/zfs/dmu_recv.c b/module/zfs/dmu_recv.c
index e534540cb5..c09c16967d 100644
--- a/module/zfs/dmu_recv.c
+++ b/module/zfs/dmu_recv.c
@@ -1255,7 +1255,12 @@ receive_object(struct receive_writer_arg *rwa, struct drr
                 * earlier in the stream.
                 */
                txg_wait_synced(dmu_objset_pool(rwa->os), 0);
-               object = drro->drr_object;
+
+               if (dmu_object_info(rwa->os, drro->drr_object, NULL) != ENOENT)
+                       return (SET_ERROR(EINVAL));
+
+               /* object was freed and we are about to allocate a new one */
+               object = DMU_NEW_OBJECT;
        } else {
                /* object is free and we are about to allocate a new one */
                object = DMU_NEW_OBJECT;
```

I'll see about getting a proper PR open for review.

**Comment by zrav** [comment]
I tested with zfs-recv-realloc and I can confirm this fixes the "dataset does not exist" issue. Thanks a lot, this was causing us trouble for our backups.
The occasional pathological receive performance seems to be unaffected though, which looks like write iops amplification (thousands of write iops/s while the transfer speed is down to <10KB/s for up to hours), which is an issue on spinning media. I will open another issue when I have more details.

**Comment by jwittlincohen** [comment]
@zrav Your second issue sounds like #8458.

**Comment by riyad** [comment]
FWIW: I was seeing this error every once in a while when backing up my laptop (Ubuntu 19.04; ZFS 0.7.12). Sometimes it would go away when skipping a few snapshots other times I had to create a new dataset and receive into that. It stopped happening after updating to ZFS 0.8.1.
